nodeletctl/pkg/nodeletctl/nodeletctl.go

```diff
@@ -106,6 +106,130 @@ func CreateCluster(cfgPath string) error {
 	return nil
 }
 
+func UpgradeCluster(cfgPath string) error {
+
+	clusterCfg, err := ParseBootstrapConfig(cfgPath)
+	if err != nil {
+		zap.S().Infof("Failed to Parse Cluster Config: %s", err)
+		return fmt.Errorf("Failed to Parse Cluster Config: %s", err)
+	}
+
+	if clusterCfg.MasterVipEnabled {
+		if clusterCfg.MasterVipInterface == "" {
+			return fmt.Errorf("MasterVipInterface cannot be empty when MasterVip is enabled")
+		}
+		if clusterCfg.MasterVipVrouterId < 1 || clusterCfg.MasterVipVrouterId > 255 {
+			return fmt.Errorf("MasterVipVrouterId should be in the range 1-255 when MasterVip is enabled")
+		}
+	}
+
+	// TODO: Need to validate if nodelet version in the bootstrap is the supported version, possibly by
+	// checking a manifest file in the rpm/deb package
+
+	if err != nil {
+		return fmt.Errorf("failed to DetermineKubePkgName: %s", err)
+	}
+
+	globalClusterStatus = new(ClusterStatus)
+	globalClusterStatus.statusMap = make(map[string]*NodeStatus)
+
+	var masterList = make(map[string]string)
+	for _, host := range clusterCfg.MasterNodes {
+		if host.NodeIP != nil {
+			masterList[host.NodeName] = *host.NodeIP
+		} else {
+			masterList[host.NodeName] = host.NodeName
+		}
+	}
+
+	masters, err := GetCurrentMasters(clusterCfg)
+	if err != nil {
+		return fmt.Errorf("Failed to get active K8s masters: %s", err)
+	}
+
+	workers, err := GetCurrentWorkers(clusterCfg)
+	if err != nil {
+		return fmt.Errorf("Failed to get active K8s workers: %s", err)
+	}
+
+	if len(clusterCfg.MasterNodes) != len(masters) {
+		return fmt.Errorf("Masters in the cluster and Bootstrap config doesnt match", err)
+	}
+
+	if len(clusterCfg.WorkerNodes) != len(workers) {
```

Comment:
Same here, checking just length doesnt tell us much, if we want validtion then let's also check the set of hosts are equal